enhancements/machine-config/mco-network-configuration.md

```diff
@@ -0,0 +1,253 @@
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The ability to configure additional host networking parameters is paramount on some platforms, especially baremetal. Baremetal nodes usally have many interfaces that need to be dynamically configured to access other, critical datacenter resources. Doing so with MachineConfig would incur reboots that are expensive in time on the baremetal platform. The ability to configure host networking parameters without rebooting is desirable. It is also desirable to rollback automatically on failure, as opposed to failing on reboot.
+
+## Motivation
+
+With hybrid clouds, node-networking setup is becoming even more challenging.
+Different payloads have different networking requirements, and not everything
+can be satisfied as overlays on top of the main interface of the node (e.g.
+SR-IOV, L2, other L2).
+
+The [Container Network Interface](https://github.com/containernetworking/cni) (CNI) standard enables different solutions for connecting networks on the node with pods. Some of them are [part of the standard](https://github.com/containernetworking/plugins), and there are others that extend support for [Open vSwitch bridges](https://github.com/kubevirt/ovs-cni), [SR-IOV](https://github.com/hustcat/sriov-cni), and more...
+
+However, in all of these cases, the node must have the networks setup before the pod is scheduled. Setting up the networks in a dynamic and heterogenous cluster, with dynamic networking requirements, is a challenge by itself - and this is what this project is addressing.
+
+The node network configuration shall remain in the MachineConfigOperator workflow, but be declarative.
+
+### Goals
+
+* Enable an admin to modify node networking configuration without rebooting (Day 2)
+* Enable cluster administrators to modify node networking after cluster installation (day 2).
+* Integration with MachineConfig node configuration for consistency, familiarity and roll out functionality.
+* Rebootless configuration without deferring errors to a later reboot
+* Rolling back cluster breaking configuration mistakes.
+* 
+### Non-Goals
+
+* Replace SRIOV operator
+
+## Proposal
+
+#### Option A
+
+The node networking configuration shall driven by MCO (and by extension kubernetes-nmstate) and executed by
+[nmstate](https://nmstate.github.io/).
+
+Do a KubeletConfig like integration:
+
+* NodeNetworkState will be introduced. This allows the user to see the network state of a node in the same declarative language that can be used to configure it.
+* The user will configure with NodeNetworkConfigurationPolicy (NNCP). This is a declaritive method of defining network state, to be applied by nmstate.
```

Comment:
We are trying to maintain the same interface as the upstream kubernetes-nmstate.

Reply:
Even using `apiVersion: nmstate.io/v1alpha1`?

But we are not actually going to be running kubernetes-nmstate?

Reply:
Definitely not the apiVersion. It would move into the machineconfig namespace.. At this time, I don't think we'll be directly running k8s-nmstate. The desire to maintain compatibility is so that those familiar with k8s-nmstate already will have an easier time adapting.